ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java

```diff
@@ -84,27 +83,32 @@
       EXPIRES_TS, DELETED_TS, NAMED_BLOBS, ACCOUNT_ID, CONTAINER_ID);
 
   /**
-   * Create a blob name to blob ID mapping if a record for that blob name does not exist.
-   * If there is an existing record but the existing blob was soft deleted or expired, allow it to be overwritten.
-   * If there is an existing record and the existing blob is still valid (not expired or soft deleted), do not
-   * overwrite it.
-   * <p>
-   * Eventually the put operation will support updates to the mapping in the third case above, but that requires other
-   * work around concurrency control.
+   * Attempt to insert a new mapping into the database.
    */
-  private static final String PUT_QUERY = String.format(
-      "INSERT INTO %1$s (%2$s, %3$s, %4$s, %5$s, %6$s) VALUES (?, ?, ?, ?, ?) ON DUPLICATE KEY UPDATE "
-          + "%5$s = IF(%8$s, VALUES(%5$s), %5$s), %6$s = IF(%8$s, VALUES(%6$s), %6$s), %7$s = IF(%8$s, null, %7$s)",
-      NAMED_BLOBS, ACCOUNT_ID, CONTAINER_ID, BLOB_NAME, BLOB_ID, EXPIRES_TS, DELETED_TS, IS_DELETED_OR_EXPIRED);
+  private static final String INSERT_QUERY =
+      String.format("INSERT INTO %s (%s, %s, %4$s, %5$s, %6$s) VALUES (?, ?, ?, ?, ?)", NAMED_BLOBS, ACCOUNT_ID,
+          CONTAINER_ID, BLOB_NAME, BLOB_ID, EXPIRES_TS);
 
-  private static final String SELECT_FOR_DELETE_QUERY =
+  /**
+   * If a record already exists for a named blob, attempt an update if the record in the DB represents an expired or
+   * deleted blob.
+   */
+  private static final String UPDATE_IF_DELETED_OR_EXPIRED_QUERY =
+      String.format("UPDATE %s SET %s = ?, %s = ?, %s = null WHERE %s AND %s", NAMED_BLOBS, BLOB_ID, EXPIRES_TS,
```

Comment:
Just curious, why are you updating the blobId column here?

Reply:
This would allow one to replace a soft deleted record with a new blob.
A put to the same key when a value was previously deleted is an allowed operation.